src/Workspaces/CoreTestUtilities/TestExportJoinableTaskContext.cs

```diff
@@ -37,23 +42,25 @@ public TestExportJoinableTaskContext()
             }
         }
 
-        private static (JoinableTaskContext joinableTaskContext, SynchronizationContext synchronizationContext) CreateJoinableTaskContext()
+        private static (JoinableTaskContext joinableTaskContext, SynchronizationContext synchronizationContext) CreateJoinableTaskContext(IDispatcherTaskJoiner? dispatcherTaskJoiner)
         {
             Thread mainThread;
             SynchronizationContext synchronizationContext;
-            if (SynchronizationContext.Current is DispatcherSynchronizationContext)
+
+            var currentContext = SynchronizationContext.Current;
+            if (currentContext != null && dispatcherTaskJoiner?.IsDispatcherSynchronizationContext(currentContext) == true)
             {
                 // The current thread is the main thread, and provides a suitable synchronization context
                 mainThread = Thread.CurrentThread;
-                synchronizationContext = SynchronizationContext.Current;
+                synchronizationContext = currentContext;
             }
             else
             {
                 // The current thread is not known to be the main thread; we have no way to know if the
```

Comment:
This needs to have a `Contract.ThrowIfTrue` that verifies the current synchronization context is _not_ a WPF synchronization context.

Reply:
Are you saying that if `currentContext` is not null then it must be the WPF dispatcher context?

Reply:
> Are you saying that if `currentContext` is not null then it must be the WPF dispatcher context?

That's not the case for all applications, but for ours yes.

Reply:
Turns out the check fails:

![image](https://github.com/dotnet/roslyn/assets/41759/e4093cc9-651a-4711-aaf9-fddfda950d3d)